src/module_3/assign_3.md

```diff
@@ -0,0 +1,980 @@
+
+    ax[1].plot(recall, precision, marker='.')
+    ax[1].set_xlabel('Recall')
+    ax[1].set_ylabel('Precision')
+    ax[1].set_title('Precision-Recall Curve (AUC-PR = {:.2f})'.format(average_precision))
+    plt.tight_layout()
+    plt.show()
+
+
+```
+
+
+```python
+#Base model with no penalties
+log_model_base = LogisticRegression(penalty= None, n_jobs=-1, class_weight='balanced').fit(train_data_x,train_data_y)
+
+y_train_pred_base = log_model_base.predict(train_data_x)
+y_train_prob_base = log_model_base.predict_proba(train_data_x)[:,1]
+
+draw_roc_prc(train_data_y, y_train_prob_base,"No regularization")
+print(metrics.confusion_matrix(train_data_y,y_train_pred_base))
+
+```
+
+
+    
+![png](assign_3_files/assign_3_12_0.png)
+    
+
+
+    [[1394040       0]
+     [  21434       0]]
+
+
+
+```python
+def best_threshold(y_true:pd.core.series.Series, y_prob:np.ndarray)->int:
+    '''
+    Function to obtain the threshold that minimizes false positive to be the closest possible to 1%.
+    '''
+    mean = y_prob.mean()
+    std = y_prob.std()
+    thresholds = np.arange(mean - 3*std, mean + 3*std, std/5)
+
+    best_threshold = None
+
+    for threshold in thresholds:
+
+        y_pred = (y_prob >= threshold).astype(int)
+        
+        false_positives = metrics.confusion_matrix(y_true, y_pred)[0][1]
+
+        if false_positives > y_true.sum()/100:
+            best_threshold = threshold
+
+    print(f"Best Threshold: {best_threshold}")
+
+    return best_threshold
+```
+
+
+```python
+threshold_base = best_threshold(train_data_y, y_train_prob_base)
+```
+
+    Best Threshold: 0.49994339123985077
+
+
+
+```python
+#Model with l2 Regularization
+
+param_grid = {
+    'C': [10 ** i for i in range(-4,3)],
+    'penalty': ['l2']}
+
+log_model_reg = LogisticRegression(n_jobs=-1, class_weight='balanced')
+
+grid_search_l2 = GridSearchCV(log_model_reg, param_grid, cv=5, scoring='accuracy',n_jobs=-1).fit(train_data_x,train_data_y)
+
+print(grid_search_l2.best_params_)
+
+y_train_pred_l2 = grid_search_l2.predict(train_data_x)
+y_train_prob_l2 = grid_search_l2.predict_proba(train_data_x)[:,1]
+
+draw_roc_prc(train_data_y, y_train_prob_l2,"With regularization")
+print(metrics.confusion_matrix(train_data_y,y_train_pred_l2))
+
+```
+
+    {'C': 0.0001, 'penalty': 'l2'}
+
+
+
+    
+![png](assign_3_files/assign_3_15_1.png)
+    
+
+
+    [[1394040       0]
+     [  21434       0]]
+
+
+
+```python
+threshold_l2 = best_threshold(train_data_y, y_train_prob_l2)
+```
+
+    Best Threshold: 0.49994339123985077
+
+
+
+```python
+#New balanced dataset with 50% 0 incomes and 1 outcomes
+train_data = df[df['order_id'].isin(train_data_orders)]
+majority_class = train_data[train_data['outcome'] == 0]
+minority_class = train_data[train_data['outcome'] == 1]
+
+majority_class_undersampled = resample(majority_class, replace=False, n_samples=len(minority_class), random_state=42)
+
+balanced_df = pd.concat([majority_class_undersampled, minority_class])
```

Comment:
As we discuss in class, balancing the train set doesn't need to be better, we are discarding a lot of negative data until we get both classes balanced. That information may be useful and have prediction power.

I'd evaluate it against unbalance case always, in my experience balancing doesn't tend to yield improvements.

Reply:
Perfect, makes sense